### PR TITLE
charge gas from SoulETH if balance is enough

### PR DESCRIPTION
**Pull request by blockchaindevsh**

### CODE REVIEW

File: core/state_transition.go [hunk 1]
```diff
@@ -241,6 +244,124 @@ func (st *StateTransition) to() common.Address {
...
+}
+
+func (st *StateTransition) SubSoulBalance(account common.Address, amount *big.Int) (err error) {
+	_, _, err = st.evm.Call(vm.AccountRef(SoulBurner), types.SoulETHAddr, burnSoulBalanceData(account, amount), callSoulGasLimit, common.Big0)
+	return
+}
+
+func (st *StateTransition) AddSoulBalance(account common.Address, amount *big.Int) {
+
+	_, _, err := st.evm.Call(vm.AccountRef(SoulMinter), types.SoulETHAddr, mintSoulBalanceData(account, amount), callSoulGasLimit, common.Big0)
```
Comment (qizhou) [inline core/state_transition.go]: Will it take extra gas for the evm.Call?

File: core/state_transition.go [hunk 2]
```diff
@@ -242,6 +245,128 @@ func (st *StateTransition) to() common.Address {
...
+
+func (st *StateTransition) SubSoulBalance(account common.Address, amount *big.Int) (err error) {
+	_, _, err = st.evm.Call(vm.AccountRef(depositorAddress), types.SoulGasTokenAddr, burnSoulBalanceData(account, amount), callSoulGasLimit, common.U2560)
+	if err == nil {
+		st.state.SubBalance(types.SoulGasTokenAddr, uint256.MustFromBig(amount))
+	}
+	return
+}
+
+func (st *StateTransition) AddSoulBalance(account common.Address, amount *big.Int) {
```
Comment (qizhou) [inline core/state_transition.go:362]: Should we check whether the soul token is backed by gas token?

File: core/state_transition.go [hunk 3]
```diff
@@ -555,8 +698,10 @@ func (st *StateTransition) innerTransitionDb() (*ExecutionResult, error) {
 		// are 0. This avoids a negative effectiveTip being applied to
 		// the coinbase when simulating calls.
 	} else {
+
 		fee := new(uint256.Int).SetUint64(st.gasUsed())
 		fee.Mul(fee, effectiveTipU256)
+		// always add to balance whether or not gasFromSoul is true
```
Comment (qizhou) [inline core/state_transition.go]: If the token is not backed by gas token, then we should just burn the token, right?

File: core/state_transition.go (hunk 1 above)
Comment (blockchaindevsh) [reply core/state_transition.go]: It's free for users  since `st.gasRemaining` is not updated.

File: core/state_transition.go (hunk 2 above)
Comment (blockchaindevsh) [reply core/state_transition.go:362]: Yeah the implementation is now assuming it's backed by gas token, I'll make it configurable in a follow-up commit.

File: core/state_transition.go (hunk 3 above)
Comment (blockchaindevsh) [reply core/state_transition.go]: Yeah the implementation is now assuming it's backed by gas token, I'll make it configurable in a follow-up commit.

File: core/state_transition.go (hunk 2 above)
Comment (qizhou) [reply core/state_transition.go:362]: I will suggest at least leaving a TODO here.

File: core/state_transition.go (hunk 3 above)
Comment (qizhou) [reply core/state_transition.go]: See the above comment.

**Review by qizhou — COMMENTED** [review]
What is your test plan for this feature?

**Comment by blockchaindevsh** [comment]
> What is your test plan for this feature?

https://github.com/blockchaindevsh/optimism/blob/c587aac95bbbb23a171073a9e8f501e184e18a9a/op-e2e/op_geth_test.go#L1028 There're now 3 basic cases, we can add more cases as necessary.

(Needs to run `make devnet-up` first)

File: core/state_transition.go (hunk 2 above)
Comment (blockchaindevsh) [reply core/state_transition.go:362]: I just pushed an implementation that honors `EnableSoulGasToken` and `IsSoulBackedByNative` option from `ChainConfig`: https://github.com/ethstorage/op-geth/pull/1/commits/3985b7a44b3c86aeb047c6ee1fca78d41882c511

File: core/state_transition.go (hunk 3 above)
Comment (blockchaindevsh) [reply core/state_transition.go]: Yes, we'll burn the `SoulGasToken` if it's not backed by gas token.

File: params/config.go [hunk 4]
```diff
@@ -439,6 +439,10 @@ type OptimismConfig struct {
 	EIP1559Elasticity        uint64 `json:"eip1559Elasticity"`
 	EIP1559Denominator       uint64 `json:"eip1559Denominator"`
 	EIP1559DenominatorCanyon uint64 `json:"eip1559DenominatorCanyon"`
+	// Enable using SoulGasToken for gas fee.
+	EnableSoulGasToken bool `json:"enableSoulGasToken"`
+	// Whether SoulGasToken is backed by native token or minted by whitelisted miners, only effective when EnableSoulGasToken is true
+	IsSoulBackedByNative bool
```
Comment (qizhou) [inline params/config.go]: Should we add `json` tag for this variable?

File: core/genesis.go [hunk 5]
```diff
@@ -296,6 +300,10 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, triedb *triedb.Database, g
 			if overrides != nil && overrides.OverrideOptimismInterop != nil {
 				config.InteropTime = overrides.OverrideOptimismInterop
 			}
+			if overrides != nil && config.IsOptimism() {
```
Comment (qizhou) [inline core/genesis.go]: What is the purpose of overrides?

File: params/config.go (hunk 4 above)
Comment (blockchaindevsh) [reply params/config.go]: Good catch, done!

File: core/genesis.go (hunk 5 above)
Comment (blockchaindevsh) [reply core/genesis.go]: After further thought, it seems `EnableSoulGasToken`/`IsSoulBackedByNative` don't have to be override-able, they can only be configured in genesis now.

File: eth/backend.go [hunk 6]
```diff
@@ -233,6 +233,8 @@ func New(stack *node.Node, config *ethconfig.Config) (*Ethereum, error) {
 		overrides.OverrideOptimismInterop = config.OverrideOptimismInterop
 	}
 	overrides.ApplySuperchainUpgrades = config.ApplySuperchainUpgrades
+	overrides.EnableSoulGasToken = config.EnableSoulGasToken
```
Comment (qizhou) [inline eth/backend.go]: I guess we also need to remove these two lines?

**Review by qizhou — APPROVED** [review]

File: eth/backend.go (hunk 6 above)
Comment (blockchaindevsh) [reply eth/backend.go]: Fixed.

**Review by qizhou — APPROVED** [review]

File: core/state_transition.go [hunk 7]
```diff
@@ -273,21 +403,44 @@ func (st *StateTransition) buyGas() error {
...
-	if have, want := st.state.GetBalance(st.msg.From), balanceCheckU256; have.Cmp(want) < 0 {
-		return fmt.Errorf("%w: address %v have %v want %v", ErrInsufficientFunds, st.msg.From.Hex(), have, want)
+
+	st.gasFromSoul = false
+	if st.evm.ChainConfig().IsOptimism() && st.evm.ChainConfig().Optimism.UseSoulGasToken {
+		have, err := st.GetSoulBalance(st.msg.From)
+		if err != nil {
+			return fmt.Errorf("GetSoulBalance error:%v", err)
+		}
+		if have, want := have.ToBig(), new(big.Int).Sub(balanceCheck, st.msg.Value); have.Cmp(want) >= 0 {
```
Comment (qzhodl) [inline core/state_transition.go:420]: Do we support the case where the soul gas token < gas needed, the gas token < gas needed, but the soul gas token + gas token > gas needed?

Comment (blockchaindevsh) [reply core/state_transition.go:420]: Nope, we don't support partially paying gas with soul gas token.

**Comment by blockchaindevsh** [comment]
Close since already merged into op-es branch.

**Pull request closed without merging**